src/audio/drc/drc_math_generic.c

```diff
@@ -0,0 +1,270 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2020 Google LLC. All rights reserved.
+//
+// Author: Pin-chih Lin <johnylin@google.com>
+
+#include <sof/audio/drc/drc_math.h>
+#include <sof/audio/format.h>
+#include <sof/math/decibels.h>
+#include <sof/math/numbers.h>
+#include <math.h>
+
+/* Add definitions for fixed-point math function usage */
+#define DRC_FIXED_DB2LIN
+#define DRC_FIXED_LIN2DB
+#define DRC_FIXED_LOG10
+#define DRC_FIXED_LOGF
+#define DRC_FIXED_SIN
+#define DRC_FIXED_ASIN
+#define DRC_FIXED_POW
+#define DRC_FIXED_EXP
+
+#define DRC_PI_FLOAT 3.141592653589793f
+#define DRC_PI_OVER_TWO_FLOAT 1.57079632679489661923f
+#define DRC_TWO_OVER_PI_FLOAT 0.63661977236758134f
+#define DRC_ONE_OVER_SQRT2 0.70710678118654752f
+
+inline float decibels_to_linear(float decibels)
+{
+#ifdef DRC_FIXED_DB2LIN
+/*
+ * Input is Q8.24: max 128.0
+ * Output is Q12.20: max 2048.0
+ */
+	return Q_CONVERT_QTOF(db2lin_fixed(Q_CONVERT_FLOAT(decibels, 24)), 20);
+#else
+	/* 10^(x/20) = e^(x * log(10^(1/20))) */
+	return expf(0.1151292546497022f * decibels);
+#endif
+}
+
+static inline float warp_log10f(float x)
+{
+#ifdef DRC_FIXED_LOG10
+/*
+ * Input is Q6.26: max 32.0
+ * Output range ~ (-inf, 1.505); regulated to Q6.26: (-32.0, 32.0)
+ */
+#define q_v 26
+#define q_mult(a, b, qa, qb, qy) ((int32_t)Q_MULTSR_32X32((int64_t)a, b, qa, qb, qy))
+	/* Coefficients obtained from:
+	 * fpminimax(log10(x), 5, [|SG...|], [1/2;sqrt(2)/2], absolute);
+	 * max err ~= 6.088e-8
+	 */
+	const int32_t HALF_Q_V = Q_CONVERT_FLOAT(0.5f, q_v);
+	const int32_t ONE_Q_V = Q_CONVERT_FLOAT(1.0f, q_v);
+	const int32_t ONE_OVER_SQRT2 = Q_CONVERT_FLOAT(DRC_ONE_OVER_SQRT2, q_v);
+	const int32_t A5 = Q_CONVERT_FLOAT(1.131880283355712890625f, q_v);
+	const int32_t A4 = Q_CONVERT_FLOAT(-4.258677959442138671875f, q_v);
+	const int32_t A3 = Q_CONVERT_FLOAT(6.81631565093994140625f, q_v);
+	const int32_t A2 = Q_CONVERT_FLOAT(-6.1185703277587890625f, q_v);
+	const int32_t A1 = Q_CONVERT_FLOAT(3.6505267620086669921875f, q_v);
+	const int32_t A0 = Q_CONVERT_FLOAT(-1.217894077301025390625f, q_v);
+	const int32_t LOG10_2 = Q_CONVERT_FLOAT(0.301029995663981195214f, q_v);
+	int32_t x_fixed = Q_CONVERT_FLOAT(x, q_v);
+	int32_t e = 0; /* Q31.1 */
+	int32_t x2, x4; /* Q2.30 */
+	int32_t A5Xx, A3Xx; /* Q6.26 */
+
+	/* frexpf implementation */
```

Comment:
What's the worst case convergance cost here ? i.e. can it loop significantly and consume too many cycles leading to XRUNs in the pipeline ? I'm assuimng here it will only loop max 32 times ?